[Dynamic Dashboard] Handle hiding Blaze card when not available

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #11359 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds logic to hide the Blaze card when not available, it depends on the changes in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2995

I was planning on handling the Onboarding on the same PR, but given the need for the FluxC changes, I decided to split them.

### Testing instructions
#### Private site
1. Make a WooExpress site private.
2. Open the app.
3. Confirm the Blaze card is hidden from both the dashboard and its settings.

#### Site with published products
1. Use a site that doesn't have any published products.
2. Open the app.
3. Confirm the Blaze card is hidden.
4. Add a new product.
5. Go back to the dashboard.
6. Confirm the Blaze card was added.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Support different texts for the badge of unavailable cards

hichamboushaba committed:
Remove the coroutines dispatcher override

We already supply the correct dispatcher when creating the DataStore, using the IO dispatcher for the other operations will just exhaust threads allocated for the IO dispatcher for no need, and might even cause deadlocks.

hichamboushaba committed:
Rename variable

hichamboushaba committed:
Hide Blaze card when unavailable

hichamboushaba committed:
Bump fluxc

hichamboushaba committed:
Hide the blaze card when the store doesn't have any published products

hichamboushaba committed:
Improve the logic of the blaze card

This avoids unnecessary fetch of products as we will handle it in the repository now

hichamboushaba committed:
Extract logic of the availabilities to external classes

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Add a TODO comment about code cleanup

dangermattic commented:
<!--
  0 Errors
  2 Warnings: Class <code>ObserveBlazeWidget...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ObserveBlazeWidgetStatus</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ObserveStatsWidgetsStatus</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Merge branch 'trunk' into issue/11359-hide-onboarding-blaze

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/data/DashboardRepository.kt:
```diff
@@ -1,84 +1,67 @@
+        .stateIn(
+            scope = siteCoroutineScope,
+            started = SharingStarted.Lazily,
+            initialValue = DashboardWidget.Status.Unavailable(R.string.my_store_widget_unavailable)
```
This is a change from the previous behavior, on the past we started by showing the stats and then hide them if we find out that the site doesn't have any orders, I changed it to this because when we don't have any orders cached, we'll always use a network call to detect if the site has orders or not, unlike the other case where we'll use just a DB call.

On the first case, and since it takes some time, before this change, the cards gets removed from the dashboard abruptly if the site doesn't have any orders, but now it's the other way around, the cards will be added, and the interaction doesn't seem to me too bad, and this will happen only once when the orders are not cached.

Please let me know what you think.

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/data/DashboardRepository.kt:
```diff
@@ -1,84 +1,67 @@
+        WooLog.e(WooLog.T.DASHBOARD, "Failed to update dashboard data", throwable)
     }
 
-    suspend fun updateWidgetVisibility(type: DashboardWidget.Type, isVisible: Boolean) = withContext(dispatchers.io) {
```
Just curious, why did you remove the IO dispatcher switch here?

hichamboushaba replied:
Sorry, I was planning on adding a comment about it, but forgot about it, the [commit](https://github.com/woocommerce/woocommerce-android/commit/0e9ad3425e5d529a8c230f8ca2637e093ee07cc6) has the reason, which is that the DataStore already uses the IO dispatcher, that we pass to its factory, so there was no need for wrapping the whole operation in it, and also given the fact that the IO dispatcher uses a limited number of threads, overusing them for non blocking operations could lead to deadlocks in some occasions.

0nko replied:
Ha, you're right! I should know, I wrote the factory injection.. 😅 I'm curious and impressed, did you just realize it when you saw this?

hichamboushaba replied:
> I'm curious and impressed, did you just realize it when you saw this?

😅 It was something that I read about in the past, it was about Room and not DataStore, but it's the same thing since Room uses its own Dispatcher as well.

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/data/DashboardRepository.kt:
```diff
@@ -1,84 +1,67 @@
-        }.onFailure { throwable ->
-            WooLog.e(WooLog.T.DASHBOARD, "Failed to update dashboard data", throwable)
-        }
+    suspend fun updateWidgets(widgets: List<DashboardWidget>) = runCatching {
```
NP: The `DashboardDataStore.updateDashboard()` already contains a similar log with `runCatching`.

hichamboushaba replied:
Good point, I'll remove it, just for context, I didn't add it, I just removed the dispatchers override here 🙂

0nko reviewed (commented):
Nice work! 👍 I tested the PR and it works as expected. 

Just one observation: I already had some products cached in the app for a store. Then I deleted all products on the web switched to that store in the app. I could still see the Blaze card because the products didn't get refreshed until I tapped on the Products tab. Do you think we should refresh products in the My Store tab, as well?

hichamboushaba commented:
> Just one observation: I already had some products cached in the app for a store. Then I deleted all products on the web switched to that store in the app. I could still see the Blaze card because the products didn't get refreshed until I tapped on the Products tab. Do you think we should refresh products in the My Store tab, as well?

Thanks @0nko for the review, regarding this point, in a previous PR, I updated the logic of refreshing products in the Dashboard to be only when we don't have any cached products, and during a pull-to-refresh, this was done to reduce the number of requests the app does on launch, and I discussed it with Jorge [here](https://github.com/woocommerce/woocommerce-android/pull/11320#discussion_r1570393898), for a store to go from having products to none is an edge case IMO.
WDYT about this? I don't have a strong opinion about it, and we might still need to change when working on iteration 2 as we'll have some other cards.

0nko commented:
@hichamboushaba Thanks for your response!

> WDYT about this? I don't have a strong opinion about it, and we might still need to change when working on iteration 2 as we'll have some other cards.

Yeah, no problem, it's an edge case, let's keep it as it is. I was just wondering about it when I noticed it.

hichamboushaba committed:
Remove redundant runCatching

hichamboushaba committed:
Bump fluxc

0nko reviewed (approved):
:shipit:

hichamboushaba committed:
Merge branch 'trunk' into issue/11359-hide-onboarding-blaze

# Conflicts:
#	build.gradle
